src/main/java/com/cburch/logisim/std/io/extra/Buzzer.java

```diff
@@ -302,44 +312,58 @@ public Object clone() {
     }
 
     public void ThreadFunc() {
-      AudioFormat af = new AudioFormat(SAMPLE_RATE, 16, 2, true, false);
+      AudioFormat af = null;
       Clip clip = null;
       AudioInputStream ais = null;
       int oldfreq = -1;
-      int oldch = -1;
       int oldpw = -1;
-      double oldvol = Double.NaN;
-      BuzzerWaveform oldwf = null;
       try {
         while (is_on.get()) {
-          if (hz != oldfreq || vol != oldvol || wf != oldwf || channels != oldch || (pw != oldpw && wf == BuzzerWaveform.Square)) {
+          if (updateRequired) {
+            updateRequired = false;
+
             if (!(hz >= 20 && hz <= 20000)) {
               return;
             }
 
-            int phase = 0;
-            if (clip != null) {
-              int oldfw = SAMPLE_RATE / oldfreq;
-              int oldphase = (int) (clip.getLongFramePosition() % (2 * oldfw));
-              var delta =
-                  SAMPLE_RATE * Math.asin(BuzzerWaveform.Sine.strategy.amplitude(oldphase, oldfreq, oldpw / 256.0))
-                      / (Math.PI * hz);
-              if (Math.abs(oldphase - oldfw) > (oldfw / 2)) {
-                phase = (int) (oldphase - delta);
-              } else {
-                phase = (int) (oldphase + delta - SAMPLE_RATE / hz);
+            if (hz != oldfreq)
+            {
+              sampleRate = (int)Math.ceil(44100.0 / hz) * hz;
+              af = new AudioFormat(sampleRate, 16, 2, true, false);
+              oldfreq = hz;
+            }
+
+            int cycle = Math.max(1, sampleRate / hz);
+            double[] values = new double[4 * cycle];
+            for (int i = 0; i < values.length; i++)
+            {
+              values[i] = wf.strategy.amplitude(i / (double)sampleRate, hz, pw / 256.0);
```

Comment:
Looking at this code, it leaves the impression that the waveform is calculated over and over again when its parameters change. In addition, these calculations are done using floating point. This may cause performance problems when simulating more complex circuits that use one or more buzzer components.

A more efficient way to approach this may be to work with look-up tables that store pre-calculated values of the different waveform kinds. By iterating through them with a changing step size, different frequencies can be obtained. The starting index for the iteration allows adjusting the wave. The amplitude can be obtained by multiplying the value read from the LUT. By using fix-point representation for the sample amplitudes and scaling factor, multiplication gets cheaper. This approach is called [direct digital synthesis](https://www.analog.com/media/en/training-seminars/tutorials/MT-085.pdf) and is often used in arbitrary waveform generators, but can be also adopted in software to calculate the waveform samples, you hand over at a fixed sampling rate to your sound card.

Reply:
The logic was inherited from the original buzzer code, which only supported sine waves and, worse than that, computed the values in real time as long as the buzzer was active.

Here, I'm only calculating 2 time periods worth of data, and I'm only doing it once, when the parameters are modified.

Using LUTs would be an even bigger performance improvement, though I'm not sure right now of how to implement that without making the code a whole lot more complex. I'll put it in my to-do list though; maybe for another PR?

Reply:
Thanks for your reply @zdimension. As I wrote: Implementing this kind of optimisation is not urgent and can be done at a later point. One could leave a comment in the code that this optimisation potential exists.